File: 공소연/programmers/P42576.js
// 링크: https://school.programmers.co.kr/learn/courses/30/lessons/42576?language=javascript
// 시간: 43.65ms
// 메모리: 68.3MB

function solution(participant, completion) {
    // reduce
    const cmap = completion.reduce((acc, cur) => {
        acc[cur] = (acc[cur] ? acc[cur] + 1 : 1);
        return acc;
    }, {})
    
    // using falsy
    return participant.find(cur => {
        if (cmap[cur]) cmap[cur] = cmap[cur] - 1;
        else return cur;
    })
}

File: 공소연/programmers/P42576.java
// 링크: https://school.programmers.co.kr/learn/courses/30/lessons/42576?language=java
// 시간: 89.88ms
// 메모리: 97.6MB

import java.util.*;

class Solution {
    public String solution(String[] participant, String[] completion) {
        String answer = "";
        
        // one not in completion but in participant
        // max nlogn
        Map<String, Integer> completionMap = new HashMap<>();
        for (String com : completion) {
            completionMap.put(com, completionMap.getOrDefault(com, 0) + 1);
        }
        
        for (String par : participant) {
            if (completionMap.getOrDefault(par, 0) == 0) {
                answer = par;
                break;
            } else {
                completionMap.put(par, completionMap.get(par) - 1);
            }
        }
        return answer;
    }
}
